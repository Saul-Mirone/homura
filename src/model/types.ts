export interface Source {
  id: number;
  name: string;
  link: string;
  sourceUrl: string;
  icon?: string;
}

export type ISOString = string;

export interface Post {
  id: number;
  sourceId: number;
  guid: string;
  link: string;
  title: string;
  content: string;
  unread: boolean;
  starred: boolean;
  date: ISOString;
}

export type PostStatus = 'unread' | 'starred';
