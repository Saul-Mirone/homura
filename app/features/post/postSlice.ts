import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { channel } from '../../channel/child';
import type { AppThunk, RootState } from '../../store';

type State = {
  content: string;
};

const initialState: State = {
  content: '',
};

const postSlice = createSlice({
  name: 'post',
  initialState,
  reducers: {
    setContent: (state, action: PayloadAction<string>) => {
      state.content = action.payload;
    },
  },
});

export const { setContent } = postSlice.actions;

export const getPostContentById = (postId: number): AppThunk => async (
  dispatch
) => {
  const post = await channel.getPostById(postId);
  dispatch(setContent(post.content));
};

export const postReducer = postSlice.reducer;

export const selectPost = (state: RootState) => {
  const { activeId, posts } = state.list;
  const { content } = state.post;

  if (!activeId)
    return {
      activeId,
    };

  const target = posts.find((x) => x.id === activeId);

  if (!target)
    return {
      activeId,
    };

  const { title, sourceName, unread, starred, date } = target;

  const post = {
    title,
    sourceName,
    unread,
    date,
    starred,
    content,
  };

  return {
    activeId,
    post,
  };
};
