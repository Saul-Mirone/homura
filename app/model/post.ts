import {
  Association,
  BelongsToCreateAssociationMixin,
  BelongsToGetAssociationMixin,
  BelongsToSetAssociationMixin,
  DataTypes,
  Model,
  Optional,
  Sequelize,
} from 'sequelize';
import type { Source } from './source';

export interface PostAttributes {
  id: number;
  sourceId: number;
  guid: string;
  title: string;
  content: string;
  unread: boolean;
  starred: boolean;
  date: Date;
}

export type CreatePostAttributes = Optional<
  PostAttributes,
  'id' | 'starred' | 'unread'
>;

export class Post
  extends Model<PostAttributes, CreatePostAttributes>
  implements PostAttributes {
  public id!: number;

  public sourceId!: number;

  public guid!: string;

  public title!: string;

  public content!: string;

  public unread!: boolean;

  public starred!: boolean;

  public date!: Date;

  public toJSON!: () => PostAttributes;

  public getSource!: BelongsToGetAssociationMixin<Source>;

  public setSource!: BelongsToSetAssociationMixin<Source, number>;

  public createSource!: BelongsToCreateAssociationMixin<Source>;

  public static associations: {
    source: Association<Post, Source>;
  };
}

export const initPost = (sequelize: Sequelize): void => {
  Post.init(
    {
      id: {
        type: DataTypes.INTEGER,
        autoIncrement: true,
        primaryKey: true,
      },
      sourceId: {
        type: DataTypes.INTEGER,
        allowNull: false,
      },
      guid: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      title: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      content: {
        type: DataTypes.TEXT,
        allowNull: false,
      },
      unread: {
        type: DataTypes.BOOLEAN,
        defaultValue: true,
      },
      starred: {
        type: DataTypes.BOOLEAN,
        defaultValue: false,
      },
      date: {
        type: DataTypes.DATE,
        allowNull: false,
      },
    },
    {
      sequelize,
      tableName: 'posts',
    }
  );
};
