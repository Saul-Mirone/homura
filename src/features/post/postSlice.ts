import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { DateTime } from 'luxon';
import { channel } from '../../channel/child';
import { format } from '../../constants/Date';
import type { AppThunk, RootState } from '../../store';

type State = {
  content: string;
};

const initialState: State = {
  content: '',
};

const postSlice = createSlice({
  name: 'post',
  initialState,
  reducers: {
    setContent: (state, action: PayloadAction<string>) => {
      state.content = action.payload;
    },
  },
});

export const { setContent } = postSlice.actions;

export const getPostContentById = (postId: number): AppThunk => async (
  dispatch
) => {
  const post = await channel.getPostById(postId);
  dispatch(setContent(post.content));
};

export const postReducer = postSlice.reducer;

export const selectPost = (state: RootState) => {
  const { activeId, posts } = state.list;
  const { content } = state.post;

  if (!activeId)
    return {
      activeId,
    };

  const target = posts.find((x) => x.id === activeId);

  if (!target)
    return {
      activeId,
    };

  const { title, sourceName, unread, starred, date, link } = target;

  const post = {
    title,
    sourceName,
    unread,
    link,
    date: DateTime.fromISO(date).toFormat(format),
    starred,
    content,
  };

  return {
    activeId,
    post,
  };
};
