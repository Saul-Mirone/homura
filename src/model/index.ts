import Database from 'better-sqlite3';
import { getPost } from './getPost';
import { getPostList, GetPostListOptions } from './getPostList';
import { getSourceList } from './getSourceList';
import { initialize } from './initialize';
import { markPostsAsRead } from './markPostsAsRead';
import { subscribe, SubscribePayload } from './subscribe';
import { SourceToPayload, sync } from './sync';
import { PostStatus } from './types';
import { unsubscribe } from './unsubscribe';
import { updatePostStatus, UpdatePostStatusOptions } from './updatePostStatus';
import { updateSource, UpdateSourceOptions } from './updateSource';

export class Model {
  private readonly sqlite: Database.Database;

  constructor() {
    const dbPath = ':memory:';
    this.sqlite = new Database(dbPath, { verbose: console.log });
    this.initialize();
  }

  private initialize() {
    return initialize(this.sqlite);
  }

  public subscribe(payload: SubscribePayload) {
    return subscribe(this.sqlite, payload);
  }

  public unsubscribe(id: number) {
    return unsubscribe(this.sqlite, id);
  }

  public getSourceList(type: PostStatus) {
    return getSourceList(this.sqlite, type);
  }

  public getPostList(options: GetPostListOptions) {
    return getPostList(this.sqlite, options);
  }

  public updateSource(options: UpdateSourceOptions) {
    return updateSource(this.sqlite, options);
  }

  public markPostsAsRead(sourceId?: number) {
    return markPostsAsRead(this.sqlite, sourceId);
  }

  public updatePostStatus(options: UpdatePostStatusOptions) {
    return updatePostStatus(this.sqlite, options);
  }

  public getPost(id: number) {
    return getPost(this.sqlite, id);
  }

  public sync(fn: SourceToPayload) {
    return sync(this.sqlite, fn);
  }
}
