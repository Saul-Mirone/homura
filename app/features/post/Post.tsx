import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Reader } from '../../components/Reader';
import { Toolkit } from '../../components/Reader/Toolkit';
import { AppDispatch } from '../../store';
import { markActiveStarredAs, markActiveUnreadAs } from '../list/listSlice';
import { getPostContentById, selectPost } from './postSlice';

export const Post: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const { post, activeId } = useSelector(selectPost);

  React.useEffect(() => {
    if (!activeId) return;
    dispatch(getPostContentById(activeId));
  }, [activeId, dispatch]);

  const toolkit = post ? (
    <Toolkit
      starred={post.starred}
      onSwitchStarred={(x) => dispatch(markActiveStarredAs(!x))}
      unread={post.unread}
      onSwitchUnread={(x) => dispatch(markActiveUnreadAs(!x))}
    />
  ) : null;

  return <Reader post={post} toolkit={toolkit} />;
};
