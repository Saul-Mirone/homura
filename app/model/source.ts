import {
  Association,
  DataTypes,
  HasManyAddAssociationMixin,
  HasManyCountAssociationsMixin,
  HasManyCreateAssociationMixin,
  HasManyGetAssociationsMixin,
  HasManyHasAssociationMixin,
  Model,
  Optional,
  Sequelize,
} from 'sequelize';
import { Post } from './post';

export interface SourceAttributes {
  id: number;
  name: string;
  link: string;
  sourceUrl: string;
  icon: string | null;
}

export type CreateSourceAttributes = Optional<SourceAttributes, 'id'>;

export class Source
  extends Model<SourceAttributes, CreateSourceAttributes>
  implements SourceAttributes {
  public id!: number;

  public name!: string;

  public link!: string;

  public sourceUrl!: string;

  public icon!: string | null;

  public getPosts!: HasManyGetAssociationsMixin<Post>;

  public addPost!: HasManyAddAssociationMixin<Post, number>;

  public hasPost!: HasManyHasAssociationMixin<Post, number>;

  public countPosts!: HasManyCountAssociationsMixin;

  public createPost!: HasManyCreateAssociationMixin<Post>;

  public posts?: Post[];

  public toJSON!: () => SourceAttributes & {
    posts: Array<ReturnType<Post['toJSON']>>;
  };

  public static associations: {
    posts: Association<Source, Post>;
  };
}

export const initSource = (sequelize: Sequelize): void => {
  Source.init(
    {
      id: {
        type: DataTypes.INTEGER,
        autoIncrement: true,
        primaryKey: true,
      },
      sourceUrl: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      name: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      icon: {
        type: DataTypes.STRING,
        allowNull: true,
      },
      link: {
        type: DataTypes.STRING,
        allowNull: false,
      },
    },
    { sequelize, tableName: 'sources' }
  );

  Source.hasMany(Post, {
    sourceKey: 'id',
    foreignKey: 'sourceId',
    as: 'posts',
  });
  Post.belongsTo(Source, {
    targetKey: 'id',
    foreignKey: 'sourceId',
    as: 'source',
  });
};
