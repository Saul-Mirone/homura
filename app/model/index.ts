import { FindOptions, Sequelize } from 'sequelize';
import { Preset } from '../constants/Preset';
import { CreatePostAttributes, initPost, Post, PostAttributes } from './post';
import { CreateSourceAttributes, initSource, Source } from './source';

export type SourceJSON = ReturnType<Source['toJSON']>;
export type PostJSON = ReturnType<Post['toJSON']>;

export type DiffSourceOptions = Omit<
  CreateSourceAttributes,
  'sourceUrl' | 'name'
> & {
  posts: Omit<CreatePostAttributes, 'sourceId'>[];
};

export type CreateSourceOptions = CreateSourceAttributes & {
  posts: Omit<CreatePostAttributes, 'sourceId'>[];
};

export type CreateSourceResult = Omit<SourceJSON, 'posts'> & {
  posts: Array<Omit<PostJSON, 'content'>>;
};

export type GetSourceListResult = Array<
  Omit<SourceJSON, 'posts'> & { count: number }
>;

export class DB {
  private readonly sequelize: Sequelize;

  private initialized: boolean;

  constructor() {
    this.sequelize = new Sequelize('sqlite::memory:');
    this.initialized = false;
  }

  public async init(): Promise<void> {
    initPost(this.sequelize);
    initSource(this.sequelize);
    await this.sequelize.sync();
    this.initialized = true;
  }

  public async createSource({
    posts,
    ...sourceOptions
  }: CreateSourceOptions): Promise<CreateSourceResult> {
    this.checkInitialized();
    const source = await Source.create(sourceOptions);
    await Promise.all(posts.map((post) => source.createPost(post)));

    const result = await Source.findByPk(source.id, {
      include: [
        {
          model: Source.associations.posts.target,
          as: Source.associations.posts.as,
          attributes: ['id', 'title', 'unread', 'starred', 'date'],
        },
      ],
      rejectOnEmpty: true,
    });

    return result.toJSON();
  }

  public async diffWithSource(id: number, options: DiffSourceOptions) {
    this.checkInitialized();
    const { posts, ...rest } = options;
    await Source.update(rest, {
      where: {
        id,
      },
    });
    const prevSource = await Source.findByPk(id, {
      include: [
        {
          model: Source.associations.posts.target,
          as: Source.associations.posts.as,
          attributes: ['id', 'title', 'unread', 'starred', 'date', 'guid'],
        },
      ],
      rejectOnEmpty: true,
    });
    let prevPosts = [...(await prevSource.getPosts())];

    await Promise.all(
      posts.map(async (post) => {
        const existPost = await Post.findOne({
          where: {
            sourceId: id,
            guid: post.guid,
          },
        });
        if (existPost) {
          await Post.update(
            {
              title: post.title,
              content: post.content,
              date: post.date,
            },
            {
              where: {
                id: existPost.id,
              },
            }
          );
          prevPosts = prevPosts.filter((p) => p.guid === post.guid);
          return;
        }
        await Post.create({
          title: post.title,
          sourceId: id,
          content: post.content,
          date: post.date,
          guid: post.guid,
        });
      })
    );

    await Promise.all(
      prevPosts.map((x) =>
        Post.destroy({
          where: {
            guid: x.guid,
          },
        })
      )
    );
  }

  public async updateSourceNameById(id: number, name: string): Promise<void> {
    this.checkInitialized();
    await Source.update(
      { name },
      {
        where: {
          id,
        },
      }
    );
  }

  public async getSourceList(
    count: 'unread' | 'starred'
  ): Promise<GetSourceListResult> {
    this.checkInitialized();
    const sourceList = await Source.findAll();

    return Promise.all(
      sourceList.map(async (source) => {
        const options =
          count === 'starred'
            ? {
                starred: true,
              }
            : {
                unread: true,
              };
        const sourceCount = await source.countPosts({
          where: options,
        });
        const json = source.toJSON();
        return {
          ...json,
          count: sourceCount,
        };
      })
    );
  }

  public async getSourceUrlList() {
    this.checkInitialized();
    const sourceList = await Source.findAll();
    return sourceList.map(({ id, sourceUrl }) => ({ id, sourceUrl }));
  }

  public async getSourceById(id: number): Promise<CreateSourceResult> {
    this.checkInitialized();
    const source = await Source.findByPk(id, {
      include: [Source.associations.posts],
      rejectOnEmpty: true,
    });

    return source.toJSON();
  }

  public getPostByPreset(
    preset: Preset
  ): Promise<Array<PostJSON & { sourceName: string; icon: string | null }>> {
    this.checkInitialized();
    const getResult = async (options?: FindOptions<PostAttributes>) => {
      const posts = await Post.findAll(options);
      return Promise.all(
        posts.map(async (x) => {
          const source = await x.getSource();
          return {
            sourceName: source.name,
            icon: source.icon,
            ...x.toJSON(),
          };
        })
      );
    };
    switch (preset) {
      case Preset.Unread:
        return getResult({ where: { unread: true } });
      case Preset.Starred:
        return getResult({ where: { starred: true } });
      case Preset.Archive:
        return getResult({ where: { unread: false } });
      case Preset.All:
      default:
        return getResult();
    }
  }

  public async getPostById(id: number): Promise<PostJSON> {
    this.checkInitialized();
    const post = await Post.findByPk(id, {
      rejectOnEmpty: true,
    });

    return post.toJSON();
  }

  public async updatePostById(
    id: number,
    options: Partial<Pick<PostAttributes, 'unread' | 'starred'>>
  ): Promise<void> {
    this.checkInitialized();
    await Post.update(options, {
      where: {
        id,
      },
    });
  }

  public async markAllPostsAsReadBySourceId(sourceId: number): Promise<void> {
    this.checkInitialized();
    await Post.update(
      {
        unread: false,
      },
      {
        where: {
          sourceId,
        },
      }
    );
  }

  public countBy(type?: 'unread' | 'starred'): Promise<number> {
    this.checkInitialized();
    switch (type) {
      case 'unread':
        return Post.count({ where: { unread: true } });
      case 'starred':
        return Post.count({ where: { starred: true } });
      default:
        return Post.count();
    }
  }

  public async removeSourceById(id: number): Promise<void> {
    this.checkInitialized();
    await Post.destroy({
      where: {
        sourceId: id,
      },
    });
    await Source.destroy({
      where: {
        id,
      },
    });
  }

  private checkInitialized() {
    if (!this.initialized) throw new Error('Not initialized!');
  }
}
