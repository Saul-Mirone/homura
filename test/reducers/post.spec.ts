import { AnyAction } from 'redux';

// organize-imports-ignore
import { mockChannel } from '../test-tools/mockChannel';

import { getPostContentById, postReducer } from '../../src/features/post/postSlice';
import { Post } from '../../src/model/types';
import { formatActions } from '../test-tools/formatActions';
import { mockStore } from '../test-tools/mockStore';

test('should handle initial state', () => {
    expect(postReducer(undefined, {} as AnyAction)).toMatchSnapshot();
});

describe('getPostContentById', () => {
    it('should action list match snapshot', async () => {
        const store = mockStore();
        mockChannel.getPostById.mockResolvedValue(({ content: 'test-content' } as unknown) as Post);
        await store.dispatch(getPostContentById(123));
        expect(formatActions(store)).toMatchSnapshot();
    });

    it('should handle getPostContentById.fulfilled', () => {
        expect(
            postReducer(
                {
                    content: '',
                },
                {
                    type: getPostContentById.fulfilled,
                    payload: {
                        content: 'fake-content',
                    },
                },
            ),
        ).toMatchSnapshot();
    });
});
