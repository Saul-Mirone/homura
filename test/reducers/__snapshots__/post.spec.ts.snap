// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPostContentById should action list match snapshot 1`] = `
Array [
  Object {
    "meta": Object {
      "arg": 123,
      "requestId": "fake-request-id",
      "requestStatus": "pending",
    },
    "payload": undefined,
    "type": "post/getPostContentById/pending",
  },
  Object {
    "meta": Object {
      "arg": 123,
      "requestId": "fake-request-id",
      "requestStatus": "fulfilled",
    },
    "payload": "test-content",
    "type": "post/getPostContentById/fulfilled",
  },
]
`;

exports[`getPostContentById should handle getPostContentById.fulfilled 1`] = `
Object {
  "content": Object {
    "content": "fake-content",
  },
}
`;

exports[`should handle initial state 1`] = `
Object {
  "content": "",
}
`;
